docs/design/readme.md

```diff
@@ -0,0 +1,1704 @@
+
+Rx sessions operate like sampling ports, storing one most recently received transfer internally as an instance of `DynamicBuffer` (see below). The buffer is passed over to the application upon the next call to `receive`. A sampling port is similar to a FIFO queue of depth 1 unit, so in the future this design can be trivially generalized to support variable-depth FIFO queues inside Rx session instances that are set to the depth of 1 unit by default. Shall the application fail to collect the received transfer(s) before the FIFO queue is full, the least recently received transfer is dropped, and the new one is pushed to the opposite end of the queue. Such occurrences should normally be recorded for diagnostic purposes via the appropriate statistical counters, but this feature is currently outside of the scope of this proposal.
+
+#### `DynamicBuffer`
+
+Lizards operate on raw serialized binary blobs of data rather than high-level message representations. Transmission is performed by enqueueing serialized transfers into a private transmission queue managed by the lizard, which is easy to abstract from the application. Reception is a more complicated case because it requires a lizard to return memory to the application that is owned by the lizard, requiring the latter to free it after use in a lizard-specific manner; further and more importantly, such memory may or may not be fragmented in a gather-scatter buffer. To hide the specifics of such memory management from the application, a new abstraction is introduced, represented by the class named `DynamicBuffer`.
+
+Examples of lizard-specific management of the returned memory can be found in
+<https://github.com/OpenCyphal/libcanard/blob/73d0a9cae94307038344b0d3eac2fd6dac44e139/libcanard/canard.h#L335-L338> and
+<https://github.com/OpenCyphal/libudpard/blob/417d92cf9779e28c7b96bebba41665874cc13ad0/libudpard/udpard.h#L285-L311>.
+
+The `DynamicBuffer` provides a uniform API for dealing with the Cyphal transfer payload returned by a lizard and also implements the movable/non-copyable RAII semantics for freeing the memory allocated for the buffer once the dynamic buffer instance is disposed of. The interface hides the gather-scatter nature of the buffer, providing a simplified linearized view. The definition of the class is approximately as follows:
+
+```c++
+/// The buffer is movable but not copyable because copying the contents of a buffer is considered wasteful.
+/// The buffer behaves as if it's empty if the underlying implementation is moved away.
+class DynamicBuffer final
+{
+public:
+    static constexpr std::size_t ImplementationFootprint = sizeof(void*) * 8;
```

Comment:
I feel like we're not using PMR and CETL enough such that we'll have multiple solutions to the same problem. Take a look at the following examples for how you can manage buffers using unique_ptr and PMR:

1. https://github.com/OpenCyphal/CETL/blob/main/cetlvast/suites/docs/examples/example_07_polymorphic_alloc_deleter.cpp
2. https://github.com/OpenCyphal/CETL/blob/main/cetlvast/suites/docs/examples/example_06_memory_resource_deleter.cpp

And here's how to wrap a C allocator in `memory_resource`: https://github.com/OpenCyphal/CETL/blob/efc3904c5a73550c32809051cf78e1b428271ee8/include/cetl/pmr/o1heap_memory_resource_delegate.hpp#L32

Reply:
This class is not responsible for managing the buffer memory but rather for hiding the specifics of its management from the user. The memory management is implemented by the lizards.